00XX-efficient-msms-for-non-native-pickles-verification.md

```diff
@@ -0,0 +1,483 @@
+
+
+#### Using all the circuits rows
+
+Our circuit uses all or almost all the available rows. However, it is likely we will be able to fit any of the sections of the MSM algorithm (each taking $`\approx 2^{15}`$ rows with one FFEC addition per row) into *exactly* $N = 2^{15}$ rows.
+
+The first chunk is easy --- assuming that our RAM lookup state is shared across all the $l$ or $2l$ invocations of the `fill_buckets`, there is no need for any further communication.
+
+Sharing state between second and third sub-section can be done by reusing some of the buckets as the memory bus.
+
+We will need to adjust the algorithm a bit. Note that we don't have to use `total` and `right_sum` as variables.
+Instead, `total` will go into the zero bucket, and `right_sum` can go into the bucket number $2^{k} - 1$ (the last one), and then the second loop in the MSM algorithm can be uniform without any extra single row for aggregation. And thus we are able to split it into sub-circuits 2 and 3 while maintaining common state without any variables, just with RAM lookup. The modified second part of the MSM algorithm is as follows:
+```rust
+// ... (loop with fill_buckets) ...
+// Instead of being a variable, `total` can be "stored" in buckets[0] for efficiency
+bucket[0] = -H.scale(2^k * (2^k - 1) / 2);
+// Similarly `right_sum` can be "stored" in buckets[2^k-1] for efficiency
+// We unroll the first loop iteration because the buckets[2^k-1] slot already contains
+// the correct first-iteration value.
+bucket[0] += buckets[2^k-1];
+// Done in chunk 2
+for i in 2..buckets.length()/2 {
+    buckets[2^k-1] += buckets[2^k - i];
+    buckets[0] += buckets[2^k-1];
+}
+// Done in chunk 3, state is communicated through the RAM lookup now
+for i in bucket.length/2..buckets.length() {
+    buckets[2^k-1] += buckets[2^k - i];
+    buckets[0] += buckets[2^k-1];
+}
+```
+
+Finally, to assert that the final computed `buckets[0]` value is what we expect, we need to slightly adjust the additive lookup argument to externally (by modifying lookup boundary conditions) assert that the accumulated computation result (the discrepancy) is contained in the last constraint of the last folding iteration.
+- Our additive lookup constraint has a form of like $\frac{1}{r + v}$, and we can use an alternative accumulator boundary condition $`\mathsf{acc}' = \mathsf{acc} +
+\frac{1}{r + 0} - \frac{1}{r + v_{0}}`$ embedded into the constraint, where $v_0$ is the value contained in the zero slot. In such a way enforce $v_0$ to be present at the zero address. This approach is already taken in the zkVM implementation.
+
+
+
+### Additive Lookups
+
+In the zkVM project there is already an implementation of an additive lookup algorithm.
+- We must make sure, when moving it into our project, that it does not rely heavily on pairings -- even though BN254 supports a pairing, it is better (for future compatibility with IPA Kimchi) to ignore this.
```

Comment:
These wasn't a design considerations. The consideration was that we could 'update the table state' in the style of RAM, which pairing-based lookups make difficult and expensive.